### ISSUE TITLE
Cache GraphQL schema

### ISSUE THREAD

**Issue by chillu** [silverstripe/silverstripe-graphql#130]
# Overview

As a developer building on a decoupled CMS, I can rely on fast dynamic execution of GraphQL even in larger schemas in order to keep my site stable and minimise custom workarounds

# Acceptance Criteria

 * GraphQL queries are not significantly slower with 100 type definitions than they are with 10 type definitions
 * Schema caches using standard framework cache logic (choosing best available default cache adapter)
 * Schema invalidates when a change in YAML config has been detected
 * Can be configured per schema
 * Defining inline resolvers via anonymous functions (or closures) continues to be supported

# Notes

 * Related to [serverside graphql query caching](https://github.com/silverstripe/silverstripe-graphql/issues/164)
 * Related to [granular caching through smart proxies](https://github.com/silverstripe/silverstripe-graphql/issues/183)
 * Might require a library upgrade, and/or a major release of the module
 * See http://webonyx.github.io/graphql-php/type-system/type-language/#performance-considerations and https://github.com/webonyx/graphql-php/pull/104

**Comment by chillu** [comment]
Just benchmarked a bit, adding CRUD operations for 20 subclasses of `Page` massively increase the overhead - from 0.6s (already not great) to 12.s (unacceptable). That's concerning because we'll need all those operations available in each CMS implementation once we switch `admin/pages` over to React+GraphQL. Large implementations could easily have 100+ of DataObject subclasses, leading to 300+ types and 1000+ queries/mutations.

Doing a simple, single file query:

```
query {
  readFiles(limit:1) {
    edges {
      node {
        ...on File {
          id
          title
        }
      }
    }
  }
}
```

Before (0.6s) https://blackfire.io/profiles/f194aa4a-d428-4eb1-9d98-f11b4cfae787/graph
After (12.3s): https://blackfire.io/profiles/06681c15-dd3c-48a8-9b81-6613bdaef6eb/graph

**Comment by chillu** [comment]
Upgrading from 0.8 to 0.11 ([pull request](https://github.com/silverstripe/silverstripe-graphql/pull/131)) unfortunately doesn't fix this (even though they claim parser performance improvements): https://blackfire.io/profiles/63b9f779-9570-409a-a21e-83f972a4d673/graph

Looking into [lazy types](http://webonyx.github.io/graphql-php/type-system/schema/#lazy-loading-of-types) now

**Comment by chillu** [comment]
With XDebug disabled, it's a bit faster (8.4s): https://blackfire.io/profiles/0819c68f-40d4-4f9b-b319-9a331bfe2a06/graph

With [lazy types](http://webonyx.github.io/graphql-php/type-system/schema/#lazy-loading-of-types) we're down to 5.4s: https://blackfire.io/profiles/114c16b5-16f0-4243-9226-01e2e700ae7e/graph (pull request pending).

Looking into serialising the schema now

Note that I'm testing on PHP 7.1 through Homebrew on a new MBP. All but the first result (0.6s) use `composer dump-autoload`

**Comment by chillu** [comment]
It turns out that Blackfire is skewing the result timings. Haven't gotten to the bottom of it yet. The situation isn't as dire as depicted above, but it's still not great. Adding 100 page types takes the average query from ~60ms to ~1200ms.

Ran the following

```
for i in {1..10}; do curl 'http://localhost/graphql/' -s -w "%{time_total}\n" -H 'Origin: http://localhost' -H 'Accept-Encoding: gzip, deflate, br' -H 'Accept-Language: en-US,en;q=0.9,de;q=0.8' -H 'User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/63.0.3239.132 Safari/537.36' -H 'Content-Type: application/json' -H 'Accept: */*' -H 'Referer: http://localhost/dev/graphiql' -H 'Cookie: Phpstorm-49e760e^C1e35e160-fc13-4f5f-9f34-266a75698607; PHPSESSID=auvdi6irev4n3tt15u2q0n22bt' -H 'Connection: keep-alive' --data-binary '{"query":"query {\n  readFiles(limit:1) {\n    edges {\n      node {\n        ...on File {\n          id\n          title\n        }\n      }\n    }\n  }\n}","variables":null}' -o /dev/null; done
```

With built in types only:

```
0.084562
0.054022
0.056435
0.056173
0.054416
0.054094
0.055593
0.057678
0.055826
0.052373
```

With 20 generated page types:

```
0.350688
0.307437
0.322245
0.294258
0.304279
0.299670
0.298616
0.340503
0.324080
0.317446
```

With 100 generated page types:

```
1.243542
1.185444
1.187602
1.199025
1.171526
1.167665
1.269340
1.165222
1.208697
1.263980
```

**Comment by chillu** [comment]
I've tried to cache the schema AST, which the lib [recommends](http://webonyx.github.io/graphql-php/type-system/type-language/#performance-considerations) when using the GraphQL type language rather than PHP to define your schema. The 100 generated page types make for a 9MB type language file. Caching it is a two-step process (generating the type language from a `Schema` object, creating an AST from that, then writing it into an array). On cache fetches, this array is converted back to an AST, then into a `Schema` object. Ideally we could converted a `Schema` into an AST without this intermediary step, but that doesn't seem to be possible - mainly because the `SchemaPrinter` removes stuff that can't be expressed in GraphQL's type language (and hence serialised), like resolvers.

With 100 generated page types: 0.732546s median before, 0.6622915s median after. So a slight performance improvement. But: 37MB vs. 149MB memory usage when generating the cache, and 37MB vs. 97MB memory usage with parsing an existing cache. So this is a dead end.

Experimental branch: https://github.com/open-sausages/silverstripe-graphql/commit/d78655604d97d42bed435a59de13bd3ff7dd2233

**Comment by chillu** [comment]
Aaron has written an RFC on how to (potentially) solve this at https://github.com/silverstripe/silverstripe-graphql/issues/192

**Comment by chillu** [comment]
This has been solved, see #192 for validation.
